### PR TITLE
[FIX][ENTERPRISE] Engagement Dashboard displaying incorrect data about active users

### PR DESCRIPTION
**Pull request by matheusbsilva137**
## Proposed changes (including videos or screenshots)

 - Fix sessions' and users' grouping in the Engagement Dashboard API endpoints;
 - Fix the data displayed in the charts from the "Active users", "Users by time of day" and "When is the chat busier?" sections of the Engagement Dashboard;
 - Replace label used to describe the amount of Active Users in the License section of the Info page.

## Issue(s)

[Task - ClickUp](https://app.clickup.com/t/5131xn)

## Steps to test or reproduce

- Running an Enterprise Version's instance of Rocket.Chat and logged in as an admin, open Administration > **Info**. The license card should display the amount of **Active** users (which may be limited to a given number) -- active users do not include app users, guests and deactived users.
- Running an Enterprise Version's instance of Rocket.Chat and logged in as an admin, open Administration > **Engagement Dashboard**. All the data shown in the "Active users", "Users by time of day" and "When is the chat busier?" sections must be limited to the amount of active users displayed in the Info page.

### **Amount of users in the Info page**
the `Users` label has been replaced by `Active_Users` since only the amount of **active** users is limited in the EE (the pie chart was already comparing the amount of **active** users with the maximum amount of active users, just the label's change was needed for clarity);

### **"Active users" section in the Engagement Dashboard**
 - **Previous behavior**: the API endpoint returned an object containing dates and the amount of sessions opened in each date (`users` field). Since a user can start multiple sessions in a same date, the value of the `users` field could exceed the amount of active users in the server (_incorret behavior_).
 - **New behavior**: the API endpopint now returns an object containg dates and, for each date, a list of users (`usersList`, which is a **set** of userIds) and the `users` field (size of the `usersList` field). The API endpoint now groups all sessions by date **and `userId`** and then perform one more group to sum the amount of distinct active users in a date.
_**Front-end:**_ The amount of weekly and monthly active users used to be calculated summing the amount of daily active users in an interval. However, recurring DAU could be counted multiple times (exceeding the amount of active users in the server). The amount of MAU and WAU obtained with the help of a **set** that aggregates all distinct `userId`s (related to active sessions) in a period -- in order to assure that only **distinct** users are being counted.

### **"Users by time of day" section in the Engagement Dashboard**
 - **Previous behavior**: the API endpoint returned a `users` field associated with each 1 hour interval between the beginning and end date. The `users` field was the amount of active sessions in each 1 hour interval. However, since a user can start multiple sessions in a same hour, the value of the `users` field could exceed the amount of active users in the server (_incorret behavior_).
 - **New behavior**: the API endpopint now returns list of users (`usersList`, which is a **set** of userIds) and the `users` field (size of the `usersList` field) associated with each 1 hour interval. The API endpoint now groups all distinct `userId`s that created sessions in a same range (1 hour interval) and then add the `users` field (which is the size of each range set).

### **"When is the chat busier?" (Hours) section in the Engagement Dashboard**
 - **Previous behavior**: since each bar in the chart covers a 2 hour range, the sum of two adjacent ranges was used as the `y` value of each bar. However, since a user can have one same session active for more than one hour, these cases were being counted twice. Also, if a user opened multiple sessions in a 2 hours range, these sessions wouldn't be grouped and they would be counted as if they were from distinct users.
 - **New behavior**: applies the same new approach used in the "Users by time of day" section, but using intervals (ranges) of 2 hours.

### CODE REVIEW

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js [hunk 1]
```diff
@@ -46,21 +46,23 @@ const ActiveUsersSection = () => {
...
+
+		const distributeValueOverPoints = (usersLists, currentIndex, usersList, i, T, array, prev) => {
+			const usersSet = new Set(usersList);
+			let j;
+			let k;
+			for (j = 0, k = i; j < T && k < array.length; ++k, ++j) {
 				if (k >= 0) {
-					array[k].y += value;
+					usersLists[currentIndex + j]?.forEach((userId) => usersSet.add(userId));
+					array[k].y = Math.max(usersSet.size, array[k].y);
```
Comment (matheusbsilva137) [inline ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js:61]: Hey @pierre-lehnen-rc and @tassoevan , I'm still not so sure whether WAU and MAU are being count correctly here 😵 . Any ideas?

File: ee/app/engagement-dashboard/client/components/UsersTab/BusiestChatTimesSection.js [hunk 2]
```diff
@@ -127,7 +127,7 @@ const ContentForDays = ({ displacement, onPreviousDateClick, onNextDateClick })
 	const { value: data } = useEndpointData('engagement-dashboard/users/chat-busier/weekly-data', params);
 	const values = useMemo(() => (data ? data.month.map(({ users, day, month, year }) => ({
 		users,
-		day: String(moment.utc([year, month - 1, day, 12, 0, 0]).valueOf()),
+		day: String(moment({ year, month: month - 1, day, hour: 0, minute: 0, second: 0 }).valueOf()),
```
Comment (KevLehman) [inline ee/app/engagement-dashboard/client/components/UsersTab/BusiestChatTimesSection.js:130]: Since you're setting hour/minute/second to 0, you can skip those props.

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: ee/app/engagement-dashboard/client/components/UsersTab/BusiestChatTimesSection.js (hunk 2 above)
Comment (matheusbsilva137) [reply ee/app/engagement-dashboard/client/components/UsersTab/BusiestChatTimesSection.js]: Great, thanks for the heads-up 🤗

File: app/models/server/raw/Sessions.js [hunk 3]
```diff
@@ -136,6 +145,19 @@ export class SessionsRaw extends BaseRaw {
 						day: '$day',
 						month: '$month',
 						year: '$year',
+						userId: '$userId',
+					},
+				},
+			},
+			{
```
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply app/models/server/raw/Sessions.js:152]: I don't think so 😕.

[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply app/models/server/raw/Sessions.js:152]: I don't think so 😕
The first group is used to build a **set** of userIds, while the second stage gets each set's size (that is, it gets the _amount of distinct active users_ in a given hour range), which is used to build many charts in the front-end side.

I tried to group these two stages (using a conditional `$sum` to calculate `users` while creating the `usersList`), but for that I'd need to have a new group field (`users`) accessing the value of a field that is being created (`usersList`) in the same $group (and it seems like this is not allowed, it didn't work).

File: app/theme/client/imports/general/base.css [hunk 4]
```diff
@@ -239,7 +239,7 @@ button {
 	.rc-old .main-content-flex {
 		flex: 1 0 auto !important; /* 1 */
 
-		height: auto !important; /* 1 */ 
+		height: auto !important; /* 1 */
```
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply app/theme/client/imports/general/base.css:242]: Actually, no 😬 
But since only trailing spaces are being removed, I think we can keep these changes, can't we? (or is it better to put them back there?)

File: app/models/server/raw/Sessions.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply app/models/server/raw/Sessions.js:152]: I don't think so 😕
The first stage is used to group multiple sessions from each user in a day, while the second stage builds a **set** of active userIds for that day and counts it (that is, it gets the _amount of distinct active users_ in a given hour range), which is used to build many charts in the front-end side.

I tried to group these two stages (using a conditional `$sum` to calculate `users` while creating the `usersList`). However, for that I'd need `users` to access the value of `usersList`, but `usersList` is still being created in **the same $group** and it seems like doing this is not allowed, it didn't work.

File: app/theme/client/imports/general/base.css (hunk 4 above)
[the remark this answers is not in the document]
Comment (KevLehman) [reply app/theme/client/imports/general/base.css:242]: Well, I noticed the same change appeared on my local 😬 so probably it's better to merge it with this anyways

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js [hunk 5]
```diff
@@ -47,37 +46,38 @@ const ActiveUsersSection = () => {
...
 		const prevMauValue = createPoint(-1);
 
-		for (const { users, day, month, year } of data.month) {
-			const i = moment.utc([year, month - 1, day, 0, 0, 0, 0]).diff(period.start, 'days');
-			distributeValueOverPoints(users, i, 1, dauValues, prevDauValue);
-			distributeValueOverPoints(users, i, 7, wauValues, prevWauValue);
-			distributeValueOverPoints(users, i, 30, mauValues, prevMauValue);
+		const usersListsMap = data.month.reduce((map, dayData) => {
+			const date = moment({ year: dayData.year, month: dayData.month - 1, day: dayData.day });
+			const i = date.diff(period.start, 'days');
```
Comment (KevLehman) [inline ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js:58]: what's `i`? 👀

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js [hunk 6]
```diff
@@ -12,14 +12,13 @@ import { downloadCsvAs } from '../../../../../../client/lib/download';
...
-
 	const period = useMemo(() => ({
 		start: moment().set({ hour: 0, minute: 0, second: 0, millisecond: 0 }).subtract(30, 'days'),
-		end: moment().set({ hour: 0, minute: 0, second: 0, millisecond: 0 }).subtract(1),
+		end: moment().set({ hour: 0, minute: 0, second: 0, millisecond: 0 }).subtract(1, 'days'),
 	}), []);
 
 	const params = useMemo(() => ({
-		start: period.start.clone().subtract(30, 'days').toISOString(),
+		start: period.start.clone().subtract(29, 'days').toISOString(),
```
Comment (KevLehman) [inline ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js:21]: i'm curious, why was this needed? 👀 february?

File: app/models/server/raw/Sessions.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (KevLehman) [reply app/models/server/raw/Sessions.js:152]: Well, it makes sense 😬

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js (hunk 6 above)
Comment (tassoevan) [reply ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js:21]: @KevLehman To compute the MAU for an arbitrary `x` date we need to sum over a **closed** time interval of `[x - 29, x]`, which give us 30 days.

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js (hunk 5 above)
Comment (tassoevan) [reply ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js:58]: @KevLehman I'm not sure how to properly name a "date offset over a time inverval, in days" variable. Suggestions?

File: ee/app/engagement-dashboard/client/components/UsersTab/BusiestChatTimesSection.js [hunk 7]
```diff
@@ -127,7 +127,7 @@ const ContentForDays = ({ displacement, onPreviousDateClick, onNextDateClick })
 	const { value: data } = useEndpointData('engagement-dashboard/users/chat-busier/weekly-data', params);
 	const values = useMemo(() => (data ? data.month.map(({ users, day, month, year }) => ({
 		users,
-		day: String(moment.utc([year, month - 1, day, 0, 0, 0]).valueOf()),
+		day: String(moment({ year, month: month - 1, day }).valueOf()),
```
[the remark this answers is not in the document]
Comment (matheusbsilva137) [reply ee/app/engagement-dashboard/client/components/UsersTab/BusiestChatTimesSection.js:130]: Yeah, I think so 🤔. This way we'll display the data at the same timezone used in the other charts. Otherwise, we'd have to move all the charts timezones to UTC, which I think makes everything more "consistent", BUT it could be confusing for most users, right? Since some dates could be shown 1 day earlier (compared to the dates from their timezones).

But if we're keeping this date at the web browser's timezone, I think we should remove the UTC conversion from `currentDate`, wdyt?

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js (hunk 6 above)
Comment (KevLehman) [reply ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js:21]: Ah, that makes sense 👀 thanks for sharing that

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js (hunk 5 above)
Comment (KevLehman) [reply ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js:58]: I'd go with `dateOffset` or just `day`, since it represents the day to which `dayData` belongs in the period, isn't it? 👀

**Review by tassoevan — APPROVED** [review]

File: ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js (hunk 5 above)
Comment (matheusbsilva137) [reply ee/app/engagement-dashboard/client/components/UsersTab/ActiveUsersSection.js]: Good suggestion, I've changed `i` to `dateOffset` 🙂

**Review by tassoevan — APPROVED** [review]
